File: frontend/src/lib/local/diagnostics.ts
/**
 * jq's error text as structured errors. Mirrors `playground_jq.jq.diagnostics` and the stderr
 * reading in `playground_jq.jq.binary`, for the in-browser engine.
 */

import type { ErrorKind, JqError } from '@/lib/types'

const COMPILE_ERROR =
    /jq: error: (?<message>.+?) at <[^>]+>, line (?<line>\d+)(?:, column (?<column>\d+))?:(?:\n(?<source>[^\n]*)\n(?<marker>[ \t]*\^+))?/g

const TALLY = /^jq: \d+ compile errors?$/gm

const RUNTIME_PREFIX = /^jq: error \(at [^)]*\): /

function error(
    kind: ErrorKind,
    message: string,
    line: number | null = null,
    column: number | null = null,
    end: number | null = null,
): JqError {
    return { kind, message, line, column, end_column: end }
}

/** Every compile error in jq's stderr, located where jq says. */
export function compileErrors(text: string): JqError[] {
    const found: JqError[] = []
    for (const match of text.matchAll(COMPILE_ERROR)) {
        const groups = match.groups ?? {}
        const line = Number(groups.line)
        const column = groups.column === undefined ? null : Number(groups.column)
        const marker = groups.marker
        const end =
            column !== null && marker !== undefined
                ? column + Math.max((marker.match(/\^/g) ?? []).length, 1)
                : null
        found.push(error('compile', groups.message ?? '', line, column, end))
    }
    if (found.length > 0) return found
    const cleaned = text
        .replace(TALLY, '')
        .trim()
        .replace(/^jq: error: /, '')
    return [error('compile', cleaned || text.trim())]
}

/** Outputs, errors and stderr messages from one run's streams and exit code. */
export function readStreams(stderr: string, exitCode: number): { errors: JqError[]; messages: string[] } {
    if (exitCode === 3) return { errors: compileErrors(stderr), messages: [] }
    const errors: JqError[] = []
    const messages: string[] = []
    for (const line of stderr.split('\n')) {
        if (line === '') continue
        if (line.startsWith('jq: parse error:')) errors.push(error('input', line.replace(/^jq: /, '')))
        else if (line.startsWith('jq: error'))
            errors.push(error('runtime', line.replace(RUNTIME_PREFIX, '').replace(/^jq: error: /, '')))
        else if (line.startsWith('["DEBUG:",')) {
            const parsed = JSON.parse(line) as [string, unknown]
            messages.push(`DEBUG: ${JSON.stringify(parsed[1])}`)
        } else messages.push(line)
    }
    if (exitCode !== 0 && errors.length === 0) {
        errors.push(error('runtime', messages.join('\n') || `jq exited with status ${String(exitCode)}`))
        return { errors, messages: [] }
    }
    return { errors, messages }
}

File: src/playground_jq/jq/diagnostics.py
"""Turning jq's error text into structured errors an editor can mark."""

import re

from playground_jq.jq.models import ErrorKind, JqError

#: One compile error: the message, then where jq says it is.
COMPILE_ERROR = re.compile(
    r"jq: error: (?P<message>.+?) at <[^>]+>, line (?P<line>\d+)(?:, column (?P<column>\d+))?:"
    r"(?:\n(?P<source>[^\n]*)\n(?P<marker>[ \t]*\^+))?",
)

#: jq's closing tally, which carries nothing a person needs twice.
TALLY = re.compile(r"^jq: \d+ compile errors?$", re.MULTILINE)

#: The prefix jq puts on runtime errors printed by the command line.
RUNTIME_PREFIX = re.compile(r"^jq: error \(at [^)]*\): ", re.MULTILINE)


def compile_errors(text: str) -> list[JqError]:
    """Every compile error in jq's message, with line, column and marked span where jq gives them."""
    errors: list[JqError] = []
    for match in COMPILE_ERROR.finditer(text):
        line = int(match["line"])
        column = int(match["column"]) if match["column"] else None
        end_column: int | None = None
        marker = match["marker"]
        if column is not None and marker is not None:
            end_column = column + max(marker.count("^"), 1)
        errors.append(
            JqError(kind="compile", message=match["message"], line=line, column=column, end_column=end_column)
        )
    if errors:
        return errors
    cleaned = TALLY.sub("", text).strip()
    cleaned = cleaned.removeprefix("jq: error: ").strip()
    return [JqError(kind="compile", message=cleaned or text.strip())]


def plain_error(kind: ErrorKind, text: str) -> JqError:
    """An error jq gives no location for, with jq's prefixes removed."""
    message = RUNTIME_PREFIX.sub("", text.strip())
    message = message.removeprefix("jq: error: ").strip()
    return JqError(kind=kind, message=message)
